packages/flutter/lib/src/widgets/stretch_effect.dart

```diff
@@ -0,0 +1,239 @@
+
+  /// The overscroll strength applied for the stretching effect.
+  ///
+  /// The value should be between -1.0 and 1.0 inclusive.
+  /// For horizontal axis, Positive values apply a pull from
+  /// left to right, while negative values pull from right to left.
+  final double stretchStrength;
+
+  /// The axis along which the stretching overscroll effect is applied.
+  ///
+  /// Determines the direction of the stretch, either horizontal or vertical.
+  final Axis axis;
+
+  /// The child widget that the stretching overscroll effect applies to.
+  final Widget child;
+
+  @override
+  State<_StretchOverscrollEffect> createState() => _StretchOverscrollEffectState();
+}
+
+class _StretchOverscrollEffectState extends State<_StretchOverscrollEffect> {
+  ui.FragmentShader? _fragmentShader;
+
+  /// The maximum scale multiplier applied during a stretch effect.
+  static const double maxStretchIntensity = 1.0;
+
+  /// The strength of the interpolation used for smoothing the effect.
+  static const double interpolationStrength = 0.7;
+
+  /// A no-op [ui.ImageFilter] that uses the identity matrix.
+  static final ui.ImageFilter _emptyFilter = ui.ImageFilter.matrix(Matrix4.identity().storage);
+
+  @override
+  void dispose() {
+    _fragmentShader?.dispose();
+    super.dispose();
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    _StretchEffectShader.initializeShader();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final bool isShaderNeeded = widget.stretchStrength.abs() > precisionErrorTolerance;
+
+    final ui.ImageFilter imageFilter;
+
+    if (_StretchEffectShader._initialized) {
+      _fragmentShader?.dispose();
+      _fragmentShader = _StretchEffectShader._program!.fragmentShader();
+      _fragmentShader!.setFloat(2, maxStretchIntensity);
+      if (widget.axis == Axis.vertical) {
+        _fragmentShader!.setFloat(3, 0.0);
+        _fragmentShader!.setFloat(4, widget.stretchStrength);
+      } else {
+        _fragmentShader!.setFloat(3, widget.stretchStrength);
+        _fragmentShader!.setFloat(4, 0.0);
+      }
+      _fragmentShader!.setFloat(5, interpolationStrength);
+
+      imageFilter = ui.ImageFilter.shader(_fragmentShader!);
+    } else {
+      _fragmentShader?.dispose();
+      _fragmentShader = null;
+
+      imageFilter = _emptyFilter;
+    }
+
+    return ImageFiltered(
+      imageFilter: imageFilter,
+      enabled: isShaderNeeded,
+      // A nearly-transparent pixels is used to ensure the shader gets applied,
+      // even when the child is visually transparent or has no paint operations.
+      child: CustomPaint(
+        painter: isShaderNeeded ? _StretchEffectPainter() : null,
+        child: widget.child,
+      ),
+    );
+  }
+}
+
+/// A [CustomPainter] that draws nearly transparent pixels at the four corners.
+///
+/// This ensures the fragment shader covers the entire canvas by forcing
+/// painting operations on all edges, preventing shader optimization skips.
```

Comment:
Are there existing tests that cover this case? Can you elaborate its effect in more details?